backend/cn/lib/solver.ml

```diff
@@ -1394,53 +1394,138 @@ let model_evaluator, reset_model_evaluator_state =
+      | (s, bt) :: qs' ->
+        let qs', body = alpha_rename qs' body in
+        let s, body = IT.alpha_rename s body in
+        ((s, bt) :: qs', body)
+    in
+    let qs, body = alpha_rename qs body in
+    let body_ = translate_term solver body in
+    let qs_ =
+      List.map
+        (fun (s, bt) ->
+          let name = CN_Names.var_name s in
+          let sort = translate_base_type bt in
+          (SMT.atom name, sort))
+        qs
+    in
+    SMT.forall qs_ body_
+
+
+  let translate_lc solver = function
+    | LC.T it -> translate_term solver it
+    | LC.Forall ((s, bt), body) -> translate_forall solver [ (s, bt) ] body
+
+
+  let translate_function solver f args rbt body =
+    let loc = Locations.other __LOC__ in
+    let arg_exprs = List.map (fun (s, bt) -> IT.sym_ (s, bt, loc)) args in
+    translate_forall solver args (eq_ (apply_ f arg_exprs rbt loc, body) loc)
+
+
+  let translate_functions solver =
+    let open Definition.Function in
+    List.filter_map
+      (fun (f, def) ->
+        match def.body with
+        | Rec_Def body ->
+          (* Normally this would require the relevant functions,
+             including `f`, to already have been declared. Here this
+             happens lazily (in the `Apply` case), including for `f`. *)
+          Some (translate_function solver f def.args def.return_bt body)
+        | Def _ -> None
+        | Uninterp -> None)
+      (Sym.Map.bindings solver.globals.logical_functions)
+
+
+  let translate_foralls solver assumptions =
+    List.map (translate_lc solver) (List.filter LC.is_forall assumptions)
+end
+
+let try_hard = ref false
+
 (** The main way to query the solver. *)
 let provable ~loc ~solver ~global ~assumptions ~simp_ctxt lc =
   let _ = loc in
-  (* ISD: should we use this somehow? *)
-  let s1 = { solver with globals = global } in
-  let rtrue () =
-    model_state := No_model;
-    `True
+  let _ = global in
+  let set_model smt_solver qs =
+    let defs = SMT.get_model smt_solver in
+    let model = model_evaluator solver defs in
+    model_state := Model (model, qs)
   in
   match shortcut simp_ctxt lc with
-  | `True -> rtrue ()
+  | `True ->
+    model_state := No_model;
+    `True
   | `No_shortcut lc ->
-    let { expr; qs; extra } = translate_goal s1 assumptions lc in
-    let model_from sol =
-      let defs = SMT.get_model sol in
-      let mo = model_evaluator s1 defs in
-      model_state := Model (mo, qs)
-    in
-    let nlc = SMT.bool_not expr in
-    let inc = s1.smt_solver in
-    debug_ack_command s1 (SMT.push 1);
-    debug_ack_command s1 (SMT.assume (SMT.bool_ands (nlc :: extra)));
-    let res = SMT.check inc in
-    (match res with
+    let { expr; qs; extra } = translate_goal solver assumptions lc in
+    let nexpr = SMT.bool_not expr in
+    let inc = solver.smt_solver in
+    debug_ack_command solver (SMT.push 1);
+    debug_ack_command solver (SMT.assume (SMT.bool_ands (nexpr :: extra)));
+    (match SMT.check inc with
```

Comment:
For future ease of reviewing, it would be helpful if renaming and refactoring changes could be separated (into different PRs, or at least commits) from functional ones. These lines are are much harder to check. I appreciate it is more work to write, but it also makes digging through the commit history a lot easier, and that is read a lot more than it is written.

Happy to assist with git-fu if need be.